sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.storage.blob.transfer;
+
+import android.content.Context;
+import android.util.Log;
+
+import androidx.annotation.MainThread;
+import androidx.lifecycle.LiveData;
+import androidx.lifecycle.MutableLiveData;
+
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.HashMap;
+
+import com.azure.android.storage.blob.transfer.TransferIdInfoLiveData.TransferFlags;
+
+/**
+ * A Hash table based cache, with each entry has transferId as 'key' and 'value' as reference
+ * to {@link TransferFlags}, {@link LiveData<TransferIdOrError>} and weak reference to
+ * {@link LiveData<TransferInfo>}.
+ *
+ * The LiveData pair, i.e. TransferIdOrError LiveData and the TransferInfo LiveData
+ * in a 'value' is related such that, when transferId set on TransferIdOrError LiveData
+ * then transferInfo of corresponding transfer will start streaming from TransferInfo
+ * LiveData.
+ *
+ * A cache entry will automatically be removed when the TransferInfo LiveData in
+ * that entry's value is no longer in ordinary use. The presence of weak reference to it
+ * in the cache entry will not prevent it being collected by the garbage collector.
+ *
+ * This cache is not synchronized explicitly but implicitly synchronized by enforcing its
+ * methods to be accessible only from the main thread.
+ *
+ * The cache is designed to shared across all {@link TransferClient} instances within the
+ * same application process. This enables all Observers of a transfer [e.g. Observers of
+ * upload(tid: 1), Observers of resume(tid: 1)] to listen to same TransferInfo LiveData,
+ * irrespective of the TransferClient they used to get the TransferInfo LiveData.
+ *
+ * The cache implementation assumes that the TransferIdOrError LiveData in the value
+ * do not strongly refer to the TransferInfo LiveData in the same value either
+ * directly or indirectly, since that will prevent the TransferIdOrError LiveData
+ * from being collected.
+ */
+final public class TransferIdInfoLiveDataCache {
```

Comment:
What is was the reasoning behind making this class public? It seems it is only used within this package. Is the decision related to how the GC works?

Reply:
good catch! this must be a package-private, being public is a type.